applications/FluidDynamicsApplication/custom_processes/two_fluids_inlet_process.cpp

```diff
@@ -0,0 +1,243 @@
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 2) = it_node->GetSolutionStepValue(DISTANCE, 0);
+    }
+
+    // Preparation for variational distance calculation process
+    // setting distance of inlet nodes to 0.0
+    // setting rest of distances to 1.0
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 0) = 1.0;
+    }
+
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( mrInletModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = mrInletModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 0) = -1.0e-7;
+    }
+
+    // Variational distance calculation process is executed to calculate distance from inlet
+    p_dist_proc->Execute();
+
+    // scaling the distance values such that 1.0 is reached at the inlet
+    const double scaling_factor = 1.0 / mInletRadius;
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+
+        if ( (mInletRadius - it_node->GetSolutionStepValue(DISTANCE, 0)) >= 0 ){
+            // inside the transition radius (from 1 to 0)
+            it_node->GetSolutionStepValue(DISTANCE, 0) = scaling_factor * (mInletRadius - it_node->GetSolutionStepValue(DISTANCE, 0));
+        } else {
+            // outside the transition radius
+            it_node->GetSolutionStepValue(DISTANCE, 0) = 0.0;
+        }
+    }
+
+    // saving the value of DISTANCE to the non-historical variable AUX_DISTANCE
+    VariableUtils var_utils;
+    var_utils.SaveScalarVar( DISTANCE, AUX_DISTANCE, rRootModelPart.Nodes() );
+
+    // (*) restoring the original distance field from its stored version
+    #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 0) = it_node->GetSolutionStepValue(DISTANCE, 2);
+    }
+
+    // subdividing the inlet into two sub_model_part
+    mrInletModelPart.CreateSubModelPart("fluid_1_inlet");
+    mrInletModelPart.CreateSubModelPart("fluid_2_inlet");
+    ModelPart& rWaterInlet = mrInletModelPart.GetSubModelPart("fluid_1_inlet");
+    ModelPart& rAirInlet = mrInletModelPart.GetSubModelPart("fluid_2_inlet");
+
+    // classifying nodes (no OMP parallel possible)
```

Comment:
I think that it should be possible to do this operation in parallel if you define a local list for each OpenMP thread and you reduce them at the end of the loop. We can have a closer look to this.

Reply:
Sorry, it keeps crashing in all my desperate attempts...